### PR TITLE
Add support for testing with Teaspoon.

### PR DESCRIPTION
**Pull request by abuiles**
Initial integration of teaspoon.
- bootstrap teaspoon when creating a new project
- Add unit test for models when running the model generator.

### CODE REVIEW

File: lib/generators/templates/initializers/teaspoon.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  # When defining a suite you can provide a name and a block. If the name is left blank, :default is assumed. You can
+  # omit various directives and the defaults will be used.
+  #
+  # To run a specific suite
+  #   - in the browser: http://localhost/teaspoon/[suite_name]
+  #   - from the command line: rake teaspoon suite=[suite_name]
+  config.suite do |suite|
+    suite.boot_partial = 'eak_rails'
+
+    suite.normalize_asset_path = lambda { |filename|
```
Comment (abuiles) [inline lib/generators/templates/initializers/teaspoon.rb]: Depends on monkey patching of teaspoon

File: lib/generators/templates/test/test_helper.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+//
+// For more information: http://github.com/modeset/teaspoon
+//
+// You can require javascript files here. A good place to start is by
+// requiring your application.js.
+//= require application
+//= require fake_xml_http_request
+//= require fakehr
+//= require httpRespond
+
```
Comment (abuiles) [inline lib/generators/templates/test/test_helper.js:24]: We'll also have a helpers dir.

File: lib/generators/templates/test/test_helper.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+// Manifest
+// If you'd rather require your test files manually (to control order for instance) you can disable the suite matcher in
+// the configuration and use this file as a manifest.
+//
+// For more information: http://github.com/modeset/teaspoon
+//
+// You can require javascript files here. A good place to start is by
+// requiring your application.js.
+//= require application
+//= require fake_xml_http_request
```
Comment (abuiles) [inline lib/generators/templates/test/test_helper.js:21]: For now seems like the best alternative is to keep a copy of this files in vendor.

File: lib/generators/templates/initializers/teaspoon.rb (hunk 1 above)
Comment (rwjblue) [inline lib/generators/templates/initializers/teaspoon.rb]: Monkey patch added in cdc3d18.

**Comment by abuiles** [comment]
@rjackson started some work on the README, we still need to complete that, make it more clear.

File: README.md [hunk 4]
```diff
@@ -114,6 +119,13 @@ The following is added to your `config/` directory:
 * `config/environments/test.js` test environment settings
 * `config/serializers/` where Rails and Ember serializers will go
 
+The following is added to your `test/` directory:
+
+* `test/test_helper.js` require the application for testing and
+  declare helpers.
+* `test/teaspoon_env.js` configuration directives used when running
```
Comment (bcardarella) [inline README.md]: should this be `test/teaspoon_env.rb` instead?

File: app/views/teaspoon/spec/_eak_rails.html.erb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<%= javascript_include_tag_for_teaspoon *@suite.spec_javascripts, @javascript_options %>
```
Comment (bcardarella) [inline app/views/teaspoon/spec/_eak_rails.html.erb]: Is this file necessary here? I don't like the idea of putting a test file in the load path of other environments

File: lib/ember/appkit/rails/teaspoon.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+module Ember::Appkit::Rails::Teaspoon; end
+
+if [:development, :test].include?(Rails.env.to_sym)
```
Comment (bcardarella) [inline lib/ember/appkit/rails/teaspoon.rb]: I think this conditional should be in the file above. No teaspoon objects should be created outside of the intended environments

Comment (rwjblue) [inline lib/ember/appkit/rails/teaspoon.rb]: I originally structured it this way so that the rest of EAKR does not need to be aware of the Teaspoon setup (all Teaspoon related requires/config/etc can be constrained to this one file). The only object created and available is a module named `Ember::Appkit::Rails::Teaspoon` (that is empty).

Comment (bcardarella) [inline lib/ember/appkit/rails/teaspoon.rb]: I still think it is a bad practice to expose test objects in other environments, even if they are empty.

File: app/views/teaspoon/spec/_eak_rails.html.erb (hunk 5 above)
Comment (abuiles) [inline app/views/teaspoon/spec/_eak_rails.html.erb]: I'll check if we can move this to other place (probably somewhere under lib?) this is used to to boot the tests  https://github.com/dockyard/ember-appkit-rails/pull/175/files#diff-e3678ab7d9b4c70c12db650eb42d1b64R26.

Comment (bcardarella) [inline app/views/teaspoon/spec/_eak_rails.html.erb]: Why not add it to `config/environments/test.js.erb`?

File: lib/ember/appkit/rails/teaspoon.rb (hunk 6 above)
Comment (rwjblue) [inline lib/ember/appkit/rails/teaspoon.rb]: @bcardarella - I will definitely move it if you prefer, but no test objects are created in environments other than development and test here.  Unless you object to the module itself existing, in that case I can move that inside this `if`.  I really think that the better configuration, is to allow the rest of the system to be ignorant of these special environment configs.

Comment (bcardarella) [inline lib/ember/appkit/rails/teaspoon.rb]: There are actually a lot of objects that Ruby creates under the hood even if they are not explicitly shown here. Rubygems tracks what files are included as well. Even if objects are empty it is never a good practice to expose any objects that are unnecessary to a production environment. Ruby runtimes add up quickly.

Comment (rwjblue) [inline lib/ember/appkit/rails/teaspoon.rb]: Moved in e7c287d.

File: app/views/teaspoon/spec/_eak_rails.html.erb (hunk 5 above)
Comment (abuiles) [inline app/views/teaspoon/spec/_eak_rails.html.erb]: Good idea I'll check that.

File: lib/ember/appkit/rails/teaspoon.rb (hunk 6 above)
Comment (bcardarella) [inline lib/ember/appkit/rails/teaspoon.rb]: @rjackson what are your thoughts on instead of this having a `ember-appkit-rails/test` file that will be added to the `test_helper.rb` or `spec_helper.rb` files? This avoids the conditional entirely

File: app/views/teaspoon/spec/_eak_rails.html.erb (hunk 5 above)
Comment (abuiles) [inline app/views/teaspoon/spec/_eak_rails.html.erb]: @bcardarella About putting it in config/environments/test.js.erb, that might not be the place since this is not related with the tests per se, but just telling teaspoon how to load and run every test https://github.com/modeset/teaspoon/blob/master/app/views/teaspoon/spec/runner.html.erb

Teaspoon has some other views which will anyways be under the load path, though they only work when used inside the engine.

Comment (bcardarella) [inline app/views/teaspoon/spec/_eak_rails.html.erb]: But that's just a script that runs once, right? That would also be the case with whatever is in `config/environments/test/js`

Comment (abuiles) [inline app/views/teaspoon/spec/_eak_rails.html.erb]: Correct, but there is some kind of dependency here, to load the environment we need to require the application, but teaspoon doesn't load any assets by default, it does it in ``/= require application` in `test_helper.js`, and to get that we need that boot_partial.

Also Teaspoon mounts by default in development env too, so we can always go to /teaspoon without having to change the environment (It allows to do RAILS_ENV=test ...  too)

File: lib/ember/appkit/rails/teaspoon/spec_controller.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+module Ember::Appkit::Rails::Teaspoon::SpecController
```
Comment (abuiles) [inline lib/ember/appkit/rails/teaspoon/spec_controller.rb:1]: @bcardarella  Fixed the issue with teaspoon's partial, now It will be available just to Teaspoon's controller load path. 

Let me know if this works for you.

(kudos to @nhocki for pointing me towards view_paths :+1:)

**Comment by abuiles** [comment]
@rjackson we have now unit tests for routes too, do you think it would makes sense to add also a basic integration test when scaffolding? with current structure that's something we can do easily. I personally would like to have them.

**Comment by rwjblue** [comment]
@abuiles - Yes, I agree. I think we should have tests for route, controller, model, and an integration test to tie it all together.  The tests generated by the scaffold generator should be setup to be fully functioning (meaning they work and provide good coverage out of the box).

File: lib/ember/appkit/rails/teaspoon/spec_controller.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  def self.prepended(mod)
+    mod.send(:before_filter, :extend_view_path)
+  end
+
+  def extend_view_path
+    append_view_path Ember::Appkit::Rails::Engine.root.join('lib', 'views')
```
Comment (bcardarella) [inline lib/ember/appkit/rails/teaspoon/spec_controller.rb]: This is cleaner but I don't think `lib/views` is the proper place to put this. Adding this directory will expose a `views/` directory to the Rubygems LOAD_PATH. I'd rather have this file live in `lib/ember/appkit/rails/views/`

File: lib/ember/appkit/rails.rb [hunk 9]
```diff
@@ -17,3 +17,7 @@ module Rails; end
 require 'ember/appkit/rails/sprockets'
 require 'ember/appkit/rails/template'
 require 'ember/appkit/rails/active_support'
+
+if [:development, :test].include?(Rails.env.to_sym)
```
Comment (bcardarella) [inline lib/ember/appkit/rails.rb:21]: Why does this need to be available to the `development` environment? If this a test concern only I'd rather have this requirement be part of `test/test_helper.rb`

Comment (nhocki) [inline lib/ember/appkit/rails.rb:21]: I think this is to keep it consistent with Teaspoon, which also loads on Dev under /teaspoon. Right @abuiles?--
Nicolás Hock Isaza

File: lib/ember/appkit/rails/teaspoon/spec_controller.rb (hunk 8 above)
Comment (abuiles) [inline lib/ember/appkit/rails/teaspoon/spec_controller.rb]: @bcardarella fixed.

File: lib/ember/appkit/rails.rb (hunk 9 above)
Comment (abuiles) [inline lib/ember/appkit/rails.rb:21]: @bcardarella @nhocki correct, by default teaspoon mounts on http://localhost:3000/teaspoon, you can also have the option to run test with the command line, which are by default run on development, from Teaspoon's  page:

> stay consistent with what you might see in the browser (since that's likely running in development)."

https://github.com/modeset/teaspoon#command-line-interface

**Comment by bcardarella** [comment]
I am :+1: for this. @rjackson re you good with it?

**Comment by rwjblue** [comment]
Yep. This looks good. 

We'll get more testers once merged, and we can do a quick shakedown tour before shipping 0.5.0.

**Comment by bcardarella** [comment]
@abuiles you want to do the honors?

**Comment by abuiles** [comment]
@bcardarella sure!
